### DISCUSSION TITLE
如何对FM模型进行更改可以让它的目标任务为ranking task?

### DISCUSSION THREAD

**Discussion by zyrmj0212** [RUCAIBox/RecBole#1266] [Q&A] [answered]
您好，我注意到现在context-aware model 的目标任务均为ctr预测 (classification 任务)，那请问现在recbole支持我将其改成专门去做rating prediciton (regression任务)以及 top-n recommendation （ranking 任务）吗？这种更改会很麻烦吗？

**Comment by Wicknight** [comment, marked as answer]
@zyrmj0212 您好！RecBole 支持通过与一般推荐模型一样的完整排名 (full sort) 来评估 context-aware 推荐模型以完成一个ranking task，但此时需要确保 .inter 文件不能加载任何其他上下文信息列，这种情况下模型可以用 Recall 等 ranking-based evaluation metrics 作为评测指标。您可以参考我们的文档中有关Context-aware Recommendation模型[参数设置部分](https://recbole.io/docs/user_guide/usage/running_different_models.html?highlight=note%20recbole%20also%20support#context-aware-recommendation)的说明以及issue #891。

**Reply by zyrmj0212** [reply]
除了更改评价指标/评测任务之外，我想让FM模型直接使用BPR loss以更好地支持ranking task; 这种更改Recbole也是支持的吗？

**Comment by Wicknight** [comment]
@zyrmj0212 您好，RecBole对于context-aware类型的模型需要对数据进行label的设置，只支持以Point-wise的形式进行评估，因为这涉及到负例构造的问题，详情您可以参考issue #891, 因此这里不支持直接对loss进行更换。

**Reply by zyrmj0212** [reply]
非常感谢您的回答！那如果我自己实现了context-aware模型的负例构造方法，使正负样本都有对应的label, 那之后我是不是就可以使用BPR loss了？（另外，据我了解，之前工作已经有将BPR loss和FM模型结合了。所以不太明白为什么在recbole中context-aware model的loss目前只提供一种选择）

**Comment by Wicknight** [comment]
@zyrmj0212 因为context-aware模型大多数使用场景的目标任务均为CTR任务，加上交互信息中context信息的负例构造并没有一个统一的标准，因此我们在RecBole中只提供了以binary classification形式进行的评测。当然，您可以按照您的需要自行修改我们的代码以更好地完成您ranking task的需求，但对于一个通用性较强的工具包来说，在找到一个统一标准之前，我们不应该写死这里的逻辑，这就是我们目前只提供一种binary classification形式评测的原因。希望我的回答能够解释您的问题，我们也欢迎您与我们讨论您找到的文献以及您的代码修改！

**Reply by zyrmj0212** [reply]
👍👍，谢谢您给的非常详细以及耐心的解答！
